src/v/iceberg/conversion_outcome.h

```diff
@@ -0,0 +1,91 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#pragma once
+
+#include "base/seastarx.h"
+
+#include <seastar/core/sstring.hh>
+
+namespace iceberg {
+
+/**
+ * Class representing an outcome of schema conversion. If schema validation
+ * failed the outcome will contain an error. The type is simillar to the Either
+ * type idea, it either contains a value or result.
+ */
+template<typename SchemaT>
+class conversion_outcome {
+public:
+    using type = SchemaT;
+
+    conversion_outcome(const conversion_outcome&) = delete;
+    conversion_outcome(conversion_outcome&&) = default;
+    conversion_outcome& operator=(const conversion_outcome&) = delete;
+    conversion_outcome& operator=(conversion_outcome&&) = default;
+    ~conversion_outcome() noexcept = default;
+
+    const std::exception_ptr& error() const {
+        if (error_ == nullptr) [[unlikely]] {
+            throw std::logic_error{"not an error"};
+        }
+        return error_;
+    }
+
+    bool is_error() const { return error_ != nullptr; }
+    bool is_success() const { return !error(); }
+
+    const SchemaT& value() const {
+        if (!schema_) [[unlikely]] {
+            std::rethrow_exception(error_);
+        }
+        return schema_.value();
+    }
+
+    SchemaT release_value() && {
+        if (!schema_) [[unlikely]] {
+            std::rethrow_exception(error_);
+        }
+        return std::move(*schema_);
+    }
+
+public:
+    // creates an outcome with schema
+    static conversion_outcome success(SchemaT schema) {
+        return conversion_outcome{std::move(schema)};
+    }
+    // creates an outcome that contains an error
+    static conversion_outcome error(const std::exception_ptr& err) {
+        return conversion_outcome(err);
+    }
+
+private:
+    conversion_outcome() = default;
+
+    explicit conversion_outcome(const std::exception_ptr& err)
+      : error_(err){};
+
+    explicit conversion_outcome(SchemaT schema)
+      : schema_(std::move(schema)){};
+
+    std::optional<SchemaT> schema_;
+    std::exception_ptr error_ = nullptr;
```

Comment:
should this be a variant with monostate, schema and exception ptr?

Reply:
i think this is now outdated since i changed this to use `boost::outcome`